torch/_refs/__init__.py

```diff
@@ -0,0 +1,250 @@
+    (torch.float32,),
+    (torch.float64,),
+    (torch.complex64,),
+    (torch.complex128,)
+  )
+
+  for idx, dtypes in enumerate(ordered_datatypes):
+    if a in dtypes and b in dtypes:
+      return ordered_datatypes[idx + 1][0]
+    if a in dtypes:
+      return b
+    if b in dtypes:
+      return a
+
+def _higher_type_kind(a: Optional[type], b: Optional[type]):
+  ordered_type_kinds = (
+    bool, int, float, complex
+  )
+
+  assert a is None or a in ordered_type_kinds, a
+  assert b is None or b in ordered_type_kinds, b
+
+  if a is b:
+      return a
+
+  if a is None:
+    return b
+
+  if b is None:
+    return a
+
+  for kind in ordered_type_kinds:
+    if a is kind:
+      return b
+    if b is kind:
+      return a
+
+  raise ValueError("Received unknown types!")
+
+# Returns the type promoted version of the original tensors and scalars.
+# Type promotion works by first deciding which of four type kinds to use:
+#   bool -> integer -> floating point -> complex
+# The type kind is the highest type kind among all the arguments.
+#
+# Second the particular datatype in the type kind is determined.
+# The datatypes are partially ordered as follows:
+#
+# bool -> uint8, int8 -> int16 -> int32 -> int64 ->
+#   float16, bfloat16 -> float32 -> float64 -> complex32 -> complex64 -> complex128
+#
+# The particular datatype is chosen as follows:
+#   - if no tensor is of the highest type kind, then the default datatype for that kind is chosen
+#   - if there are only scalar tensors (tensors with no dimensions) of the highest type kind, then
+#       the datatype is the highest scalar tensor datatype. If this is not uniquely determined then
+#       it is the next highest datatype.
+#   - if there are non-scalar tensors (tensors with one or more dimensions) of the highest type kind,
+#       then the datatype is the highest non-scalar tensor datatype. If this is not uniquely determined
+#       then it is the next highest datatype.
+#
+# Once the datatype is determined all tensors are cast to it and all scalars are cast to the corresponding
+#   Python datatype (bool, int, float, complex).
+# TODO: document int_to_float
+# TODO: update to support changing default dtypes
+def type_promote_elementwise(*_args, int_to_float=False):
+  args = tuple(filter(lambda x: x is not None, _args))
+
+  # Type checking
+  for arg in args:
+    assert isinstance(arg, (Number, Tensor))
+
+  # Determines datatypes for each category
+  scalar_args = filter(lambda x: isinstance(x, Number), args)
+  scalar_type_kind = reduce(
+    lambda acc, x: _higher_type_kind(acc, type(x)),
+    scalar_args,
+    bool)
+
+  scalar_tensors = filter(lambda t: isinstance(t, Tensor) and t.ndim == 0, args)
+  scalar_tensor_dtype = reduce(
+    _higher_dtype,
+    (t.dtype for t in scalar_tensors),
+    torch.bool
+  )
+  scalar_tensor_type_kind = _dtype_to_kind(scalar_tensor_dtype)
+
+  nonscalar_tensors = filter(lambda t: isinstance(t, Tensor) and t.ndim != 0, args)
```

Comment:
All the filters in this function are not v pythonic. This variable, for example, is defined here and just used once below in a generator expression. It would be better to simply do
```python
  nonscalar_tensors_dtypes = (t.dtype for t in args if isinstance(t, Tensor) and t.ndim != 0)
```

Reply:
Yeah we can get rid of some maps and filters